docs/genesis-preparation.md

```diff
@@ -0,0 +1,352 @@
+# Genesis Preparation
+
+This article describes the steps should be taken to prepare genesis block.
+
+## 1. Nodes Keys
+
+First of all, secret/public keys for the cluster's nodes should be generated. Nodes will
+use them to perform blockchain-related procedures (for examples, to sign new blocks).
+These keys should be uploaded to the nodes during cluster deployment.
+
+### 1.1. Secret Keys
+
+To generate secret keys for nodes use this command:
+
+```
+$ stack exec -- cardano-keygen generate-key --path PATH
+```
+
+where `PATH` is a path to file for generated key. Use `.key` extension for it.
+
+For exmaple, to generate keys for 7 nodes execute this command 7 times:
+
+```
+$ stack exec -- cardano-keygen generate-key --path /tmp/node1.key
+$ stack exec -- cardano-keygen generate-key --path /tmp/node2.key
+$ stack exec -- cardano-keygen generate-key --path /tmp/node3.key
+$ stack exec -- cardano-keygen generate-key --path /tmp/node4.key
+$ stack exec -- cardano-keygen generate-key --path /tmp/node5.key
+$ stack exec -- cardano-keygen generate-key --path /tmp/node6.key
+$ stack exec -- cardano-keygen generate-key --path /tmp/node7.key
+```
+
+**Responsible**: The secret key for particular node should generate a person who
+launch this node. For Mainnet launching secret keys generation should be done by
+IOHK DevOps team.
+
+### 1.2. Public Keys
+
+It's possible to read generated secret key by this command:
+
+```
+$ stack exec -- cardano-keygen read-key --path /tmp/node1.key
+```
+
+Example of response:
+
+```
+[keygen:INFO] Processing command
+[keygen:INFO] Primary: 9645e94f800a5dbcf2ea4f21c60755ca4b75ba7efe02ea20a8278d1e2ff185f1c22cd797a8c5e131a40906f89d57fcaea6c70dcdf3869e4a6a7a6a04292ba10f; address hash: 204779e68379f19ad88c3599c22015988febb4c0258f100a61346ce1
+[keygen:INFO] No wallet set
+[keygen:INFO] Keys:
+[keygen:INFO] Vss PK: 582103803acf1550b381926d748949d5bd4d8a7b7cfe1f8112f52555d49ff88ac6db77; short = vsspub:6cc4ca60
```

Comment:
In the aforementioned branch I currently don't get a full `Vss PK`.

```
[keygen:INFO] withConfiguration using custom system start time 0
[keygen:INFO] Processing command
[keygen:INFO] Primary: pub:7593aec1; address hash: 8945967c
[keygen:INFO] No wallet set
[keygen:INFO] Keys: 
[keygen:INFO] Vss PK: vsspub:b00540a2
```

Reply:
Oh, let me check!

Reply:
Will cherry-pick related fix now.

Reply:
This was fixed in a recent commit to that branch:

```
[keygen:INFO] withConfiguration using custom system start time 0
[keygen:INFO] Processing command
[keygen:INFO] Primary: 0d696e00dd62cacfa73d28709ac620ecc99fc1c5a3a0691db6445d8b27828dc37b987046e395d381032ffc32d32e86a4ff672c6a5ce03ff38ed3150598489b1a; address hash: 2ce8bb8d054b2219195b84fb543f26a7b21931dd10cb18c231a3e835
[keygen:INFO] No wallet set
[keygen:INFO] Keys: 
[keygen:INFO] Vss PK: 5821030dbef12823dad3e5929a9158651ff2ad9fe9274d66968dc4bb30232d1ac7f982; short = vsspub:02b5efb0
```